tests/experimental/datasources/test_config.py

```diff
@@ -508,3 +509,32 @@ def test_custom_sorter_serialization(
 
     for sorter_str in expected_sorter_strings:
         assert sorter_str in dumped, f"`{sorter_str}` not found in dumped json"
+
+
+@pytest.mark.unit
+@pytest.mark.parametrize(
+    "ds_w_assets",
+    [
+        p(
+            PG_COMPLEX_CONFIG_DICT["xdatasources"]["my_pg_ds"],
+            id=PG_COMPLEX_CONFIG_DICT["xdatasources"]["my_pg_ds"]["type"],
+        ),
+        p(
+            PG_COMPLEX_CONFIG_DICT["xdatasources"]["my_pandas_ds"],
+            id=PG_COMPLEX_CONFIG_DICT["xdatasources"]["my_pandas_ds"]["type"],
+        ),
+    ],
+)
+def test_type_field_always_dumped(ds_w_assets: dict):
```

Comment:
This test didn't actually catch the error so it's of questionable value.

Reply:
Is this because you are using a dictionary with the `type` field defined and checking that the type field roundtrips?
If you don't think this adds value then feel free to remove it.

Reply:
It's because the only way to reproduce the `type` field not being serialized error is when adding assets "in-code". Because that's the only place where the default `type` field value is used. And because the default is used, pydantic doesn't consider the field as having been "set".

So when we dump the model with `exclude_unset=True` it gets left out.
https://docs.pydantic.dev/usage/exporting_models/#modeldict
Note in the above docs it says `exclude_unset` defaults to `False`, but we've changed this behavior to default to `True`

When we read from config everything (including `type`) gets loaded into the `DataAsset` class constructer and so pydantic considers it `set`, that's why the `test_config.py` tests didn't catch the issue.